File: C/0025_reverse_nodes_in_k_group/0025_reverse_nodes_in_k_group.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* reverse(struct ListNode* start, struct ListNode* dummyEnd)
{
    struct ListNode *prev = dummyEnd;
    while (start != dummyEnd) {
        struct ListNode *temp = start->next;
        start->next = prev;
        prev = start;
        start = temp;
    }
    return prev;
}

struct ListNode* reverseKGroup(struct ListNode* head, int k)
{
    struct ListNode *dummyEnd = head;
    for (int i = 0; i < k; i++) {
        if (!dummyEnd) {
            return head;
        }
        dummyEnd = dummyEnd->next;
    }
    struct ListNode* newHead = reverse(head, dummyEnd);
    head->next = reverseKGroup(dummyEnd, k);
    return newHead;
}

File: Java/0025_reverse_nodes_in_k_group/0025_reverse_nodes_in_k_group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummyEnd = head;
        for (int i = 0; i < k; i++) {
            if (dummyEnd == null) {
                return head;
            }
            dummyEnd = dummyEnd.next;
        }
        ListNode newHead = reverse(head, dummyEnd);
        head.next = reverseKGroup(dummyEnd, k);
        return newHead;
    }
    
    public ListNode reverse(ListNode start, ListNode dummyEnd) {
        ListNode prev = dummyEnd;
        while (start != dummyEnd) {
            ListNode temp = start.next;
            start.next = prev;
            prev = start;
            start = temp;
        }
        return prev;
    }
}
